### PR TITLE
Fix/dont throw for nonexisting table for getMetaData and dropTable

### LINKED ISSUE #1183: NamespaceMappingKeyValueService should not throw for operations that can be performed on non-existing tables
If using the NamespaceMappingKeyValueService, sweep throws `java.lang.IllegalArgumentException: Unable to resolve full name for table reference <test_table>` for a deleted table. It should fail silently.

AbstractKeyValueServiceTest#dropTablesShouldBeIdempotent also fails with the same exception, as it cannot resolve the name a second time.

### PR DESCRIPTION
**Pull request by hsaraogi**
This fixes #1183.

This has been tested with existing internal tests.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1186)

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/AbstractTableMappingService.java [hunk 1]
```diff
@@ -57,9 +57,10 @@ public TableReference getMappedTableName(TableReference tableRef) {
             return tableMap.get().get(tableRef);
         } else {
             updateTableMap();
-            Validate.isTrue(tableMap.get().containsKey(tableRef),
-                    "Unable to resolve full name for table reference " + tableRef);
-            return tableMap.get().get(tableRef);
+            if (tableMap.get().containsKey(tableRef)) {
```
Comment (SerialVelocity) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/AbstractTableMappingService.java]: These seem unsafe... (the old code too)
We should store a local reference to `tableMap.get()` or do a single `get(tableRef)` and then check if it is `null`.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/KVTableMappingService.java [hunk 2]
```diff
@@ -103,11 +104,19 @@ public TableReference addTable(TableReference tableRef) {
...
+            return getAlreadyExistingMappedTableName(tableRef);
         }
         return TableReference.createWithEmptyNamespace(shortName);
     }
 
+    private TableReference getAlreadyExistingMappedTableName(TableReference tableRef) {
+        try {
+            return getMappedTableName(tableRef);
+        } catch (TableMappingNotFoundException ex) {
+            throw Throwables.propagate(ex);
```
Comment (SerialVelocity) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/KVTableMappingService.java]: You should continue throwing an `IllegalArgumentException`

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java [hunk 3]
```diff
@@ -66,9 +67,13 @@ public TableMappingService getTableMapper() {
...
     public void addGarbageCollectionSentinelValues(TableReference tableRef, Set<Cell> cells) {
-        delegate().addGarbageCollectionSentinelValues(
-                tableMapper.getMappedTableName(tableRef),
-                cells);
+        try {
+            delegate().addGarbageCollectionSentinelValues(
+                    tableMapper.getMappedTableName(tableRef),
+                    cells);
+        } catch (TableMappingNotFoundException e) {
+            throw Throwables.propagate(e);
```
Comment (SerialVelocity) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java]: Same here

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java [hunk 4]
```diff
@@ -96,7 +101,11 @@ protected KeyValueService delegate() {
 
     @Override
     public void delete(TableReference tableRef, Multimap<Cell, Long> keys) {
-        delegate().delete(tableMapper.getMappedTableName(tableRef), keys);
+        try {
+            delegate().delete(tableMapper.getMappedTableName(tableRef), keys);
+        } catch (TableMappingNotFoundException e) {
+            throw Throwables.propagate(e);
```
Comment (SerialVelocity) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java]: and here

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java [hunk 5]
```diff
@@ -106,27 +115,38 @@ public void dropTable(TableReference tableRef) {
...
+        return tableNames;
+    }
+
     @Override
     public Map<Cell, Value> get(TableReference tableRef, Map<Cell, Long> timestampByCell) {
-        return delegate().get(tableMapper.getMappedTableName(tableRef), timestampByCell);
+        try {
+            return delegate().get(tableMapper.getMappedTableName(tableRef), timestampByCell);
+        } catch (TableMappingNotFoundException e) {
+            throw Throwables.propagate(e);
```
Comment (SerialVelocity) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java]: here too

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java [hunk 6]
```diff
@@ -138,7 +158,11 @@ public void dropTables(Set<TableReference> tableRefs) {
     public Multimap<Cell, Long> getAllTimestamps(TableReference tableRef,
                                                  Set<Cell> keys,
                                                  long timestamp) {
-        return delegate().getAllTimestamps(tableMapper.getMappedTableName(tableRef), keys, timestamp);
+        try {
+            return delegate().getAllTimestamps(tableMapper.getMappedTableName(tableRef), keys, timestamp);
+        } catch (TableMappingNotFoundException e) {
+            throw Throwables.propagate(e);
```
Comment (SerialVelocity) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java]: Here :)

Comment (SerialVelocity) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java]: (I won't comment on the rest)

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java [hunk 7]
```diff
@@ -151,23 +175,35 @@ public void dropTables(Set<TableReference> tableRefs) {
...
+        }
     }
 
     @Override
     public byte[] getMetadataForTable(TableReference tableRef) {
-        return delegate().getMetadataForTable(tableMapper.getMappedTableName(tableRef));
+        try {
+            return delegate().getMetadataForTable(tableMapper.getMappedTableName(tableRef));
+        } catch (TableMappingNotFoundException e) { // table does not exist.
+            return new byte[0];
```
Comment (SerialVelocity) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java]: We return an empty metadata for no table?

Comment (SerialVelocity) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java]: *empty byte array

Comment (hsaraogi) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java]: Yes, ``SweepTaskRunnerImpl`` has:

```
        if (keyValueService.getMetadataForTable(tableRef).length == 0) {
            log.warn("The sweeper tried to sweep table '{}', but the table does not exist. Skipping table.", tableRef);
            return SweepResults.createEmptySweepResult(0L);
        }
```

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java [hunk 8]
```diff
@@ -106,27 +114,38 @@ public void dropTable(TableReference tableRef) {
 
     @Override
     public void dropTables(Set<TableReference> tableRefs) {
```
Comment (clockfort) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java]: it seems like dropTables() should also drop the associated overflow tables too

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java (hunk 7 above)
Comment (clockfort) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/TableRemappingKeyValueService.java]: AtlasDbConstants.EMPTY_TABLE_METADATA

**Comment by SerialVelocity** [comment]
Apart from no tests, looks good

**Comment by jboreiko** [comment]
tests, yall.  TESTS

This is implicitly tested by sweep but really is part of the expected contract for a KVS method.  Therefor it should be added to a KVS test suite to ensure all KVS adhere to whatever assumption sweep is operating on.

**Pull request merged**